Support for Extended Tasks

jcheng5 opened the pull request:
This PR adds a couple of new features (that we're also adding to Shiny for R):

1. A `ui.input_task_button` that is similar to `ui.input_action_button`, except it automatically disables itself and shows "Processing..." after it clicks; and automatically(-ish) resets itself back to normal when the work it causes is done.
2. An `@extended_task` decorator that can be used to take a long-running task (expressed using an async function) and let it continue without blocking any reactive processing Shiny wants to do; but also makes it possible to take the results of such a computation and seamlessly integrate it back into reactive contexts, so it can be used from reactive outputs, calcs, etc.

Fixes #837 (or at least attempts to capture the spirit of it).

jcheng5 commented:
@schloerke @karangattu Either of you have any idea of why the CI playwright-examples tasks are failing?

A lot of it seems to be "UserWarning: Tight layout not applied. The left and right margins cannot be made large enough to accommodate all axes decorations." being emitted, but I don't see a way to know which particular example apps it was running when that happened.

karangattu commented:
> @schloerke @karangattu Either of you have any idea of why the CI playwright-examples tasks are failing?
> 
> A lot of it seems to be "UserWarning: Tight layout not applied. The left and right margins cannot be made large enough to accommodate all axes decorations." being emitted, but I don't see a way to know which particular example apps it was running when that happened.

Update your branch...we made a few changes to example tests. I suspect your branch does not have those changes yet.

schloerke commented:
@karangattu I believe it has the latest given the GHA runs on a merged commit. 

Looks like one error is coming from the cli create test. And the other is from layout_columns api example.

gadenbuie commented on shiny/api-examples/extended_task/app.py:
```diff
@@ -0,0 +1,50 @@
+        ui.input_task_button("btn", "Compute, slowly")
+        ui.input_action_button("btn_cancel", "Cancel")
+
+    slow_compute.bind_task_button("btn")
```
Alternatively (see other comment about where `bind_task_button()` lives)

```suggestion
    ui.bind_task_button("btn", slow_compute)
```

gadenbuie reviewed (approved):
This looks great! All of my comments are rather minor, except for one detail about whether `bind_task_button()` should be a method on `ExtendedTask` (which is a little different from how this ended up in Shiny for R).

jcheng5 commented on CHANGELOG.md:
Ugggh I keep hitting this! Thank you

jcheng5 commented on shiny/reactive/_extended_task.py:
```diff
@@ -0,0 +1,301 @@
+            # Will never get here, but make type checker happy
+            return self.value.get()
+
+    def bind_task_button(self, id: str) -> None:
```
Yeah, I realize the approach I took here is not consistent with what we decided for R. I thought this was a little different for two reasons:

1. In R, we have separate bslib and shiny packages which made the distance between the binding logic and the button it's coupled to, feel more egregious.
2. In Python, `bind_task_button(task, "btn")` feels very weird for some reason.

That being said, I don't love `task.bind_task_button()` either. I was debating adding a `task_button_id` argument to the decorator. I realize this is moving even further in the opposite direction of what we decided for R.

gadenbuie replied:
Reacting to `bind_task_button(task, "btn")` feeling weird: maybe binding to the task could be done through an additional decorator that applies to the ExtendedTask?

```suggestion
@ui.bind_task_button("btn")
@reactive.extended_task
async def slow_compute(a: int, b: int) -> int:
    await asyncio.sleep(3)
    return a + b
```

jcheng5 replied:
That's not bad. Let's look at it side by side with this:
```python
@reactive.extended_task(task_button_id="btn")
async def slow_compute(a: int, b: int) -> int:
    await asyncio.sleep(3)
    return a + b
```

gadenbuie replied:
Personally, I'm pretty attached to the clean separation between an ExtendedTask and how we manifest the task state in the UI. Especially if we thought of a (hypothetical) alternative presentation like

```python
@ui.bind_task_modal()
@reactive.extended_task
async def slow_compute(a: int, b: int) -> int:
    await asyncio.sleep(3)
    return a + b
```

jcheng5 replied:
I tried it this way and didn't have a strong preference either way, so I went with it. To get it to work correctly as both function and decorator, the `id` argument has to be keyword-only. So the final formulation looks like:

```python
@ui.bind_task_button(button_id="btn")
@reactive.extended_task
async def slow_compute(a: int, b: int) -> int:
    await asyncio.sleep(3)
    return a + b
```

jcheng5 commented on js/build.ts:
```diff
@@ -61,6 +61,11 @@ const opts: Array<BuildOptions> = [
   },
+  {
+    entryPoints: { "spin/spin": "spin/spin.scss" },
+    plugins: [sassPlugin({ type: "css", sourceMap: false })],
```
It's generated regardless. It doesn't look to me like `sourceMap` is a real option, not sure where it came from--I can remove it in both places where it's used.

jcheng5 replied:
Hmmm, scratch that. It matters for the dataframe entry point; without `sourceMap: false` the source map is actually inlined into the the generated dataframe.js. With it, the .map file is generated but the inline sourcemap is gone.

jcheng5 commented:
@schloerke @karangattu Turns out these playwright failures were not a false positive, but correctly calling out broken behavior from when I updated bslib!

@gadenbuie, when I created [bslib@v0.6.1-plus-task-button](https://github.com/rstudio/bslib/tree/v0.6.1-plus-task-button), I rebased my task button changes against the v0.6.1 CRAN release. It looks like that was too far back, as the Python `layout_columns` feature seems to rely on your refactoring in bslib that made layout_columns implemented by a web component. Can you help me figure out exactly what commits I need from bslib main, to be compatible with py-shiny main?

gadenbuie commented on scripts/htmlDependencies.R:
```diff
@@ -11,7 +11,7 @@ versions <- list()
 withr::local_temp_libpaths()
 ignore <- capture.output({
-  pak::pkg_install(c("rstudio/bslib", "cran::shiny", "cran::htmltools"))
+  pak::pkg_install(c("rstudio/bslib@v0.6.1-plus-task-button", "rstudio/shiny@main", "cran::htmltools"))
```
Sorry about missing this! I think we should move away from allowing `@main` to be the ref here. Instead, if we're not using the version on CRAN, I propose we have a moving `py-shiny` tag in these repos and that we update those tags and before running this script.

Note that the actual ref used is stored in a JSON file when the dependencies are updated.

```suggestion
  pak::pkg_install(c("rstudio/bslib@py-shiny", "rstudio/shiny@py-shiny", "cran::htmltools"))
```

I'll add a tag for shiny and bslib. I'll point shiny at HEAD on main and I'll point bslib's tag at the previously used version of bslib so you can start from there.

gadenbuie replied:
I guess there's a chicken-and-egg problem here: If we update the tag in bslib before merging in py-shiny, then for a period the tag is out of date. If we want to be careful, as in this PR, we can start the PR with a branch or short ref and then, before merging move the tag and update this script to use the new tag.

jcheng5 replied:
I'm taking the careful approach. 👍

gadenbuie commented on shiny/www/shared/bslib/_version.json:
```diff
@@ -1,5 +1,5 @@
 {
   "note!": "This file is auto-generated by scripts/htmlDependencies.R",
   "package": "bslib",
-  "version": "Github (rstudio/bslib@f05bd23d7df5a7465e418a5794925dacdd27bb6b)"
```
This is version of bslib we previously used in the last released version of py-shiny.

gadenbuie replied:
Now tagged at https://github.com/rstudio/bslib/releases/tag/py-shiny

gadenbuie commented on shiny/www/shared/_version.json:
```diff
@@ -1,5 +1,5 @@
   "note!": "This file is auto-generated by scripts/htmlDependencies.R",
   "package": "shiny",
-  "version": "CRAN (R 4.3.1)"
+  "version": "Github (rstudio/shiny@54988c17c8fae0e74e69b3b7fd99edaa435dc3c6)"
```
This is now tagged at https://github.com/rstudio/shiny/releases/tag/py-shiny
